src/providers/ws.rs

```diff
@@ -0,0 +1,497 @@
+use std::{
+    collections::{
+        btree_map::Entry,
+        BTreeMap,
+    },
+    fmt::Debug,
+};
+
+use async_trait::async_trait;
+use base64::{
+    engine::general_purpose,
+    Engine,
+};
+use ethers::types::H160;
+use futures::{
+    future,
+    select,
+    stream::Fuse,
+    Sink,
+    SinkExt,
+    Stream,
+    StreamExt,
+    TryStreamExt,
+};
+use futures_channel::mpsc;
+use tokio_tungstenite::connect_async;
+use tracing::{
+    debug,
+    error,
+    trace,
+    warn,
+};
+use tungstenite::{
+    client::IntoClientRequest,
+    Message,
+};
+use uuid::Uuid;
+
+use crate::core::{
+    error::{
+        Error,
+        Result,
+    },
+    provider::{
+        EthereumProvider,
+        Provider,
+        StreamResponse,
+        UniswapV2Provider,
+        UniswapV3Provider,
+    },
+    types::{
+        ethereum::BlockHeader,
+        query::QueryOptions,
+        uniswap_v2,
+        uniswap_v3,
+    },
+};
+
+const WS_PATH: &str = "websocket";
+
+type WsMsg = Result<tungstenite::Message, tungstenite::Error>;
+type WsResult = Result<Vec<u8>>;
+type OperationMsg = (Uuid, Operation, mpsc::UnboundedSender<WsResult>);
+
+#[derive(Clone, Debug)]
+pub struct WsProviderImpl {
+    operations: mpsc::UnboundedSender<OperationMsg>,
+}
+
+impl WsProviderImpl {
+    async fn request<T>(&self, operation: Operation) -> StreamResponse<T>
+    where
+        T: serde::de::DeserializeOwned + 'static,
+    {
+        let raw_data_stream = self.raw_request(operation).await?;
+        let raw_data_stream = raw_data_stream.into_async_read();
+
+        let stream = csv_async::AsyncDeserializer::from_reader(raw_data_stream);
+        let records = stream
+            .into_deserialize::<T>()
+            .map(|r| {
+                match r {
+                    Ok(r) => Ok(r),
+                    Err(e) => Err(Error::from(e)),
+                }
+            })
```

Comment:
This could be simplified
```suggestion
            .map_err(Error::from)
```